iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+// Replaces each usage of an entry point with its original symbol name with a
+// new symbol name.
+void replaceEntryPointUses(mlir::ModuleOp moduleOp,
+                           const DenseMap<Attribute, Attribute> &replacements) {
+  for (auto funcOp : moduleOp.getOps<mlir::FuncOp>()) {
+    funcOp.walk([&](DispatchOp dispatchOp) {
+      auto it = replacements.find(dispatchOp.entry_point());
+      if (it != replacements.end()) {
+        dispatchOp.entry_pointAttr(it->second.cast<SymbolRefAttr>());
+      }
+    });
+  }
+}
+
+bool areRegionsEquivalent(Region *lhs, Region *rhs) {
+  if (lhs->getBlocks().size() != rhs->getBlocks().size()) {
+    return false;
+  }
+
+  auto blockPairs = llvm::zip(lhs->getBlocks(), rhs->getBlocks());
+  for (auto blockPair : blockPairs) {
+    auto &lhsBlock = std::get<0>(blockPair);
+    auto &rhsBlock = std::get<1>(blockPair);
+    if (lhsBlock.getOperations().size() != rhsBlock.getOperations().size()) {
```

Comment:
You probably don't want to do this, `.size()` is O(N).

Reply:
good catch - any good LLVM-style recommendations? maybe llvm::zip_longest (as we are zipping below, so no use iterating once to count and then again to compare when we could just try comparing and if one iterator runs out (Optional None values start appearing) we know the sizes differ

Reply:
`llvm::zip_longest` uses `llvm::Optional`, which requires visibility into the destructor:

```
[build] D:\dev\projects\iree\third_party\llvm-project\llvm\include\llvm/ADT/Optional.h(65): error C2248: 'mlir::Operation::~Operation': cannot access private member declared in class 'mlir::Operation'
[build] D:\dev\projects\iree\third_party\llvm-project\mlir\include\mlir/IR/Operation.h(739): note: compiler has generated 'mlir::Operation::~Operation' here
[build] D:\dev\projects\iree\third_party\llvm-project\mlir\include\mlir/Pass/PassInstrumentation.h(17): note: see declaration of 'mlir::Operation'
[build] D:\dev\projects\iree\third_party\llvm-project\llvm\include\llvm/ADT/Optional.h(63): note: while compiling class template member function 'void llvm::optional_detail::OptionalStorage<T,false>::reset(void) noexcept'
[build]         with
[build]         [
[build]             T=mlir::Operation
[build]         ]
```

I could switch to managing iterators explicitly, but that would be pretty messy to read through. The O(N) early check doesn't seem too bad to me.

Reply:
Add a note if you don't want to do it yet, but it's precisely the kind of thing that'll death-by-a-thousand-papercuts passes like this and where we will fall over when scaling (bert is tiny and already has 1000 executables to dedupe with hundreds of instructions each - it's only going to get 10x larger - writing a helper function that takes two operation lists and compares them by tracking two iterators in 5 lines of code doesn't seem like a high cost to never have to worry about this again).

Reply:
Added a note.